pennylane/templates/subroutines/qdrift.py

```diff
@@ -0,0 +1,248 @@
+
+    Args:
+        coeffs (Array): the coefficients of the operations from each term in the hamiltonian.
+        ops (List[~.Operator]): the normalized operations from each term in the hamiltonian.
+        time (complex): time to evolve under the target hamiltonian.
+        n (int): number of samples in the product. Defaults to 1.
+        seed (int): random seed. Defaults to None.
+
+    Returns:
+        List[~.Operator]: the decomposition of operations as per the approximation.
+    """
+    normalization_factor = qml.math.sum(qml.math.abs(coeffs))
+    probs = qml.math.abs(coeffs) / normalization_factor
+    exps = [
+        qml.exp(base, qml.math.sign(coeff) * normalization_factor * time * 1j / n)
+        for base, coeff in zip(ops, coeffs)
+    ]
+
+    choice_rng = qml.math.random.default_rng(seed)
+    return choice_rng.choice(exps, p=probs, size=n, replace=True)
+
+
+class QDrift(Operation):
+    r"""An operation representing the QDrift approximation for the complex matrix exponential
+    of a given Hamiltonian.
+
+    # TODO: link paper reference, mention that we assume each operator is normalized.
+
+    The QDrift subroutine provides a method to approximate the matrix exponential of a Hamiltonian
+    expressed as a linear combination of terms which in general do not commute. For the Hamiltonian
+    :math:`H = \Sigma_j h_j H_{j}`, the product formula is constructed by random sampling from the
+    terms of the Hamiltonian with the probability :math:`p_j = h_j / \sum_{j} hj` as:
+
+    .. math::
+
+        \prod_{j}^{n} e^{i \lambda H_j \tau / n},
+
+    where :math:`\tau` is time, :math:`\lambda = \sum_j |h_j|` and :math:`n` is the total number of
+    terms to be added to the product.
+
+    Args:
+        hamiltonian (Union[~.Hamiltonian, ~.Sum]): The Hamiltonian written in terms of products of
+            Pauli gates
+        time (int or float): The time of evolution, namely the parameter :math:`t` in :math:`e^{-iHt}`
+        n (int): An integer representing the number of exponentiated terms
+        seed (int): The seed for the random number generator
+
+    Raises:
+        TypeError: The ``hamiltonian`` is not of type :class:`~.Hamiltonian`, or :class:`~.Sum`
+        QuantumFunctionError: If the coefficients of ``hamiltonian`` are trainable and are used
+            in a differentiable workflow.
+
+    **Example**
+
+    .. code-block:: python3
+
+        coeffs = [0.25, 0.75]
+        ops = [qml.PauliX(0), qml.PauliZ(0)]
+        H = qml.dot(coeffs, ops)
+
+        dev = qml.device("default.qubit", wires=2)
+        @qml.qnode(dev)
+        def my_circ():
+            # Prepare some state
+            qml.Hadamard(0)
+
+            # Evolve according to H
+            qml.QDrift(H, time=1.2, n = 10)
+
+            # Measure some quantity
+            return qml.probs()
+
+    >>> my_circ()
+    [0.71061676 0.         0.28938324 0.        ]
+
+
+    .. details::
+        :title: Usage Details
```

Comment:
Please add a note saying that differentiation wrt H coefficients is not allowed.